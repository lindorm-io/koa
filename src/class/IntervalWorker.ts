import Timeout = NodeJS.Timeout;
import { EventEmitter } from "events";
import { Logger } from "@lindorm-io/winston";
import { TObject, TPromise } from "@lindorm-io/core";

export interface IIntervalWorkerOptions {
  callback: TPromise<any>;
  time: number;
  logger: Logger;
}

enum IntervalWorkerEvent {
  START = "INTERVAL_WORKER_START",
  STOP = "INTERVAL_WORKER_STOP",
  SUCCESS = "INTERVAL_WORKER_SUCCESS",
  ERROR = "INTERVAL_WORKER_ERROR",
}

export class IntervalWorker extends EventEmitter {
  private callback: TPromise<any>;
  private interval: Timeout;
  private logger: Logger;
  private time: number;

  constructor(options: IIntervalWorkerOptions) {
    super();

    this.callback = options.callback;
    this.interval = undefined;
    this.time = options.time;

    this.logger = options.logger.createChildLogger("interval-worker");
  }

  public trigger(): void {
    this.logger.info("worker trigger");
    this.callback()
      .then((result: any) => {
        this.logger.debug("worker success", { result });
        super.emit(IntervalWorkerEvent.SUCCESS, { result });
      })
      .catch((err: Error) => {
        this.logger.error("worker error", err);
        super.emit(IntervalWorkerEvent.ERROR, err);
      });
  }

  public start(): void {
    this.logger.info("worker start", { intervalMs: this.time });
    this.interval = setInterval(() => this.trigger(), this.time);
    super.emit(IntervalWorkerEvent.START);
  }

  public stop(): void {
    this.logger.info("worker stop");
    clearInterval(this.interval);
    this.interval = undefined;
    super.emit(IntervalWorkerEvent.STOP);
  }

  public static get Event(): TObject<IntervalWorkerEvent> {
    return IntervalWorkerEvent;
  }
}
