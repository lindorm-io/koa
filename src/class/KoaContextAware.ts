import { IKoaAppAgent, IKoaAppContext, IKoaAppMetadata } from "../typing";
import { Logger } from "@lindorm-io/winston";

export abstract class KoaContextAware {
  protected agent: IKoaAppAgent;
  protected cache: unknown;
  protected handler: unknown;
  protected issuer: unknown;
  protected keystore: unknown;
  protected logger: Logger;
  protected metadata: IKoaAppMetadata;
  protected metrics: Record<string, number>;
  protected repository: unknown;
  protected token: unknown;

  constructor(ctx: IKoaAppContext) {
    this.agent = ctx.agent;
    this.cache = ctx.cache;
    this.handler = ctx.handler;
    this.issuer = ctx.issuer;
    this.keystore = ctx.keystore;
    this.logger = ctx.logger;
    this.metadata = ctx.metadata;
    this.metrics = ctx.metrics;
    this.repository = ctx.repository;
    this.token = ctx.token;
  }
}
