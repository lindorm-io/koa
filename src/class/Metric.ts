import { KoaContext } from "../typing";

export class Metric {
  private readonly ctx: KoaContext;
  private readonly key: string;
  private readonly start: number;
  private readonly current: number;

  public constructor(ctx: KoaContext, key: string) {
    this.ctx = ctx;
    this.key = key;
    this.start = Date.now();
    this.current = ctx.metrics[key] || 0;
  }

  public get(): number {
    return Date.now() - this.start;
  }

  public end(): void {
    this.ctx.metrics[this.key] = this.current + this.get();
  }
}
