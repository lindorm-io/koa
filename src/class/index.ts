export * from "./IntervalWorker";
export * from "./KoaApp";
export * from "./KoaContextAware";
