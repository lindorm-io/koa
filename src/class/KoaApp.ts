import Koa, { Middleware } from "koa";
import Router from "koa-router";
import bodyParser from "koa-bodyparser";
import userAgent from "koa-useragent";
import { IntervalWorker } from "./IntervalWorker";
import { Logger } from "@lindorm-io/winston";
import {
  bodyCaseSwitchMiddleware,
  metadataMiddleware,
  errorMiddleware,
  metricsMiddleware,
  sessionLoggerMiddleware,
} from "../middleware";

export interface IKoaAppOptions {
  logger: Logger;
  port: number;
}

export class KoaApp {
  readonly app: Koa;
  readonly router: Router;

  private loaded: boolean;
  private logger: Logger;
  private middleware: Array<Middleware>;
  private port: number;
  private workers: Array<IntervalWorker>;

  constructor(options: IKoaAppOptions) {
    this.app = new Koa();
    this.router = new Router();

    this.loaded = false;
    this.logger = options.logger;
    this.middleware = [
      userAgent,
      bodyParser(),
      bodyCaseSwitchMiddleware,
      metadataMiddleware,
      sessionLoggerMiddleware({ logger: this.logger }),
      metricsMiddleware,
      errorMiddleware,
    ];
    this.port = options.port;
    this.workers = [];
  }

  public callback(): any {
    if (!this.loaded) {
      this.load();
    }

    return this.app.callback();
  }

  public addMiddleware(middleware: Middleware): void {
    this.middleware.push(middleware);
  }

  public addRoute(route: string, router: Router): void {
    this.router.use(route, router.routes(), router.allowedMethods());
  }

  public addWorker(worker: IntervalWorker): void {
    this.workers.push(worker);
  }

  public load(): void {
    this.loaded = true;

    this.loadMiddleware();
    this.loadRouter();
    this.loadEmitter();

    this.logger.info("app is loaded");
  }

  public async start(): Promise<void> {
    if (!this.loaded) {
      this.load();
    }

    this.listen();

    await this.waitForStartEvent();

    this.logger.info("app has started");
  }

  private listen(): void {
    this.app.listen(this.port, (): void => {
      this.logger.info(`listening on server port: ${this.port}`);
      this.app.emit("start");
    });
  }

  private loadEmitter(): void {
    this.app.on("error", (error: Error): void => {
      this.logger.error("app caught error", error);
    });

    this.app.on("start", (): void => {
      this.loadWorkers();
    });
  }

  private loadMiddleware(): void {
    for (const middleware of this.middleware) {
      this.app.use(middleware);
    }
  }

  private loadRouter(): void {
    this.app.use(this.router.routes());
    this.app.use(this.router.allowedMethods());
  }

  private loadWorkers(): void {
    for (const worker of this.workers) {
      worker.start();
      worker.trigger();
    }
  }

  private async waitForStartEvent(): Promise<void> {
    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error("app start has timed out"));
      }, 30000);
      this.app.on("start", () => {
        clearTimeout(timeout);
        resolve();
      });
    });
  }
}

export { Router };
