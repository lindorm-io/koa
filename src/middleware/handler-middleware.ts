import { IKoaAppContext } from "../typing";
import { KoaContextAware } from "../class";
import { Middleware, Next } from "koa";
import { camelCase } from "lodash";

export const handlerMiddleware = (Handler: typeof KoaContextAware, key?: string): Middleware => async (
  ctx: IKoaAppContext,
  next: Next,
): Promise<void> => {
  /*
   * Ignoring TS here since KoaContextAware needs to be abstract
   * to ensure that all input at least attempts to be unique
   */
  // @ts-ignore
  ctx.handler[camelCase(key || Handler.name)] = new Handler(ctx);

  await next();
};
