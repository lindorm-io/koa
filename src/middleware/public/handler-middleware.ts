import { KoaContext, Middleware } from "../../typing";
import { KoaContextAware } from "../../class";
import { camelCase } from "lodash";

interface Options {
  key?: string;
}

export const handlerMiddleware =
  (Handler: typeof KoaContextAware, options?: Options): Middleware<KoaContext> =>
  async (ctx, next): Promise<void> => {
    const start = Date.now();

    /*
     * Ignoring TS here since KoaContextAware needs to be abstract
     * to ensure that all input at least attempts to be unique
     */
    // @ts-ignore
    ctx.handler[camelCase(options?.key || Handler.name)] = new Handler(ctx);

    ctx.metrics.handler = (ctx.metrics.handler || 0) + (Date.now() - start);

    await next();
  };
