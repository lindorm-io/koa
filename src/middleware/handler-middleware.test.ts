import { KoaContextAware } from "../class";
import { handlerMiddleware } from "./handler-middleware";

class Test extends KoaContextAware {}

describe("controllerMiddleware", () => {
  let ctx: any;
  let next: any;

  beforeEach(() => {
    ctx = {
      agent: "agent",
      cache: "cache",
      controller: { data: "data" },
      handler: {},
      issuer: "issuer",
      keystore: "keystore",
      logger: "logger",
      metadata: "metadata",
      metrics: "metrics",
      repository: "repository",
      support: "support",
      token: "token",
    };
    next = () => Promise.resolve();
  });

  test("should set handler on context", async () => {
    await expect(handlerMiddleware(Test)(ctx, next)).resolves.toBe(undefined);
    expect(ctx).toMatchSnapshot();
  });

  test("should set and keep references", async () => {
    await expect(handlerMiddleware(Test)(ctx, next)).resolves.toBe(undefined);
    ctx.controller.data = "new-data";
    expect(ctx).toMatchSnapshot();
  });

  test("should set handler with specific key", async () => {
    await expect(handlerMiddleware(Test)(ctx, next)).resolves.toBe(undefined);
    expect(ctx).toMatchSnapshot();
  });
});
