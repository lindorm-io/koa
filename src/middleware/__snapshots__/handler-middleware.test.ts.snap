// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllerMiddleware should set and keep references 1`] = `
Object {
  "agent": "agent",
  "cache": "cache",
  "controller": Object {
    "data": "new-data",
  },
  "handler": Object {
    "test": Test {
      "agent": "agent",
      "cache": "cache",
      "handler": [Circular],
      "issuer": "issuer",
      "keystore": "keystore",
      "logger": "logger",
      "metadata": "metadata",
      "metrics": "metrics",
      "repository": "repository",
      "token": "token",
    },
  },
  "issuer": "issuer",
  "keystore": "keystore",
  "logger": "logger",
  "metadata": "metadata",
  "metrics": "metrics",
  "repository": "repository",
  "token": "token",
}
`;

exports[`controllerMiddleware should set handler on context 1`] = `
Object {
  "agent": "agent",
  "cache": "cache",
  "controller": Object {
    "data": "data",
  },
  "handler": Object {
    "test": Test {
      "agent": "agent",
      "cache": "cache",
      "handler": [Circular],
      "issuer": "issuer",
      "keystore": "keystore",
      "logger": "logger",
      "metadata": "metadata",
      "metrics": "metrics",
      "repository": "repository",
      "token": "token",
    },
  },
  "issuer": "issuer",
  "keystore": "keystore",
  "logger": "logger",
  "metadata": "metadata",
  "metrics": "metrics",
  "repository": "repository",
  "token": "token",
}
`;

exports[`controllerMiddleware should set handler with specific key 1`] = `
Object {
  "agent": "agent",
  "cache": "cache",
  "controller": Object {
    "data": "data",
  },
  "handler": Object {
    "otherKey": Test {
      "agent": "agent",
      "cache": "cache",
      "handler": [Circular],
      "issuer": "issuer",
      "keystore": "keystore",
      "logger": "logger",
      "metadata": "metadata",
      "metrics": "metrics",
      "repository": "repository",
      "token": "token",
    },
  },
  "issuer": "issuer",
  "keystore": "keystore",
  "logger": "logger",
  "metadata": "metadata",
  "metrics": "metrics",
  "repository": "repository",
  "token": "token",
}
`;
